[facebook] add support

zWolfrost opened the pull request:
Fixes #470 and #2612 (probably duplicate)
For now it supports Account Photos & Photo Albums.
The only way it can work is by making one request per post, so it's not really optimized unfortunately.

zWolfrost commented:
It looks like Facebook blocks your account for about 15 minutes when running the extractor for too many images.
It happened to me, after running the extractor and downloading 1800 of them.
Also, it only appears to be an account ban (logging out removes the block), so it's probably best if the extractor actively avoids using the imported cookies, unless requested otherwise (with proper warnings).
Please let me know your thoughts.

zWolfrost commented:
> I think photos and videos have a signature in the url, Facebook maybe can track you and ban using this info.

I don't think i understand how you would be able to ban someone using a signature in the photo url. I think the most reasonable option would just be to use the request cookies (which include your account ids and such) to account-ban you.

As i mentioned, it's not really a complete ban, it's only limited to some parts of the UI, and logging out (thus not sending the account request cookies) does remove the block, with the tradeoff that you can't view private or R-18 images.

I still don't know if being logged out in advance prevents the ban altogether. If that's the case then i think i will add a warning about that.

zWolfrost commented:
After doing some more testing i can tell that not using cookies still gets you blocked from viewing images, in the sense that you are forced to log in, and it happens much faster than when using them.
I think it's best not to use cookies unless facebook forces the user to login, and print a small warning whenever the extractor uses them. Doing this i could extract about 2400 images before getting temporarily blocked, and i think that's pretty good.

Also, I'd love to know if the extractor works for anyone else other than me, so please feel free to let me know.

AdBlocker69 commented:
Hi, I've tested your version and it seems to be working fine for pictures. I'm planning to save quite a few images from a public Facebook page and was wondering if using one of the --sleep commands could circumvent you from being blocked (or if Facebook just reacts to an arbitrary number of request, no matter the frequency). And just overall - does that mean I'm generally unable to connect to the Facebook services (like using gallery-dl with it) or does it just prevent browser/account interaction?
Let me not forget - thanks for your work. I hope this gets implemented in the official project soon. Facebook is (still) such a big platform; so having a tool like gallery-dl supporting it is pretty important (imo)...

Facebook video support would be nice too: Luckily in my case they weren't that many so I was able to download them one by one with yt-dlp... ...But they also don't support Album/Account video downloading (yet) like they do with YouTube for example.

zWolfrost commented:
Hi, thank you for your feedback.

I'm not sure if waiting to continue extracting would work, and if it did work, i have no idea for how long or after how many images the wait should start. That would require a lot of testing and unfortunately every time i get blocked i have to wait about 6 hours to try again.

To be more specific, the "block" which I'm talking about only prevents you from accessing images by their URL (the way the extractor does it), but you can still access them by using the React user interface.
That means, accessing them by clicking them, scrolling with the arrows etc., but if, for example, you reload the page while viewing one, an error pops up about you "using this feature too fast".
When not using an account, instead, you don't get the error but you get redirected to the login page (aside from that the behavior is the same though).

As far as i can tell this block is only limited to this and you can do any other thing on Facebook

About the video support, i will keep that in mind. I'm not sure how yt-dlp downloads them and i will check that out when i have time.

AdBlocker69 commented:
Thanks for the info :)
Btw, do you know what to do when let's say, you have downloaded 2400 pictures and get blocked afterwards, and want to continue downloading from the same profile; can you just continue after the 6 hours because I guess when checking for duplicates gallery-dl still does requests for the 2400 images (as it goes chronologically from newest to oldest post) or does that work differently?

zWolfrost commented:
No, I'm sorry, once you get blocked the photo page doesn't load at all (assuming you're loading it by its URL) so there is no way to know the metadata and stuff. This is the reason why I just added a way to continue the extraction from an image in the set/album instead of having to start from the beginning. Just take the photo URL and add "&setextract" to it to download the whole set from there instead of the photo alone. The user will be prompted with this URL if they get blocked while extracting

AdBlocker69 commented:
Good idea for implementing that 👍🏻 Does it only work with the prompted URL because I just tried it up front by using an image link and adding "&setextract" to it but it gave me an 'unknown command' error after downloading just the singular image?

Also, it seems like your video extraction only pulls the version without audio (the best one in the list of formats in yt-dlp but it gets merged with an audio-only version there by default)... So it would be best to either add the ffmpeg merge-by-default or have it select the "hd" format by default which has video+audio.

zWolfrost commented:
No idea why the "&setextract" feature didn't work to you, to me it works fine. Maybe you used an URL that didn't have a set id in it (without "&set=...") so the extract didn't know how to continue. For reference, the URL should be something like this:
`https://www.facebook.com/photo...fbid=...&set=...&setextract`

This is also the reason why i don't think there's a way to make an "all profile videos" extractor, they don't share a set id i can use to navigate though them all.

Good catch for the audio thing though, i wasn't wearing headphones :)
I have fixed it right now, the audio gets downloaded as well. By default they will be separate, to merge the two you'll have to let youtube-dl/yt-dlp handle the download, by adding `"videos": "ytdl"` in the facebook extractor configuration

AdBlocker69 commented:
Okay, several things I just realized by doing some trial and error 😅:
First of all, I need to use the full link for gallery-dl to detect what "set" I even want to continue downloading from; I had just used the short version (as marked blue) and wondered why it didn't do anything anymore after downloading the image the link directs to.

![ss](https://github.com/mikf/gallery-dl/assets/172994569/6a826c14-2062-4725-8b0b-fe0b98a0df77)

Secondly, I need to then put the full link into quotation marks since it otherwise, as you said, detects the text after the ampersand as a secondary command (as marked red), giving me a 'syntax error' and not processing the rest further after downloading the image the link directs to.

![ss2](https://github.com/mikf/gallery-dl/assets/172994569/618a2c43-6422-4c73-a378-f7a58490a87a)

So this is how it's to be written to avoid any errors due to the link format and command logic etc.

![ss3](https://github.com/mikf/gallery-dl/assets/172994569/5be29c4b-8704-4466-92e0-40203d153235)

Now I got it work successfully 👌🏻

AdBlocker69 commented:
No problem, that's just what happens when the outgoing situation is slightly different :)
I just like to use short links as sometimes, when taking them directly from browsing the web, they have some certain parameters in them (like the size of an image etc.) which are undesirable. So it's more or less best practice for me to take the link as 'raw' as possible to avoid any of that.
In this case the extra information given in the link was vital though...

zWolfrost commented:
There, i just had to change the matching URL pattern a little. Now it works even without including the set id. Hopefully it's the same for you.

AdBlocker69 commented:
Works 🤙🏻

lobt4 commented:
Hi, how would I go about saving each filename in this format (highlighted in blue):
![chrome_vBJ2LoC8Ux](https://github.com/user-attachments/assets/928f01f7-2b40-46f9-8b62-a9449cd2ed2e)

instead of this current default format (appended with .jpg)?
![chrome_w0icE2BLT6](https://github.com/user-attachments/assets/1e3f11e9-5cde-4661-8e67-55175a412f67)

lobt4 commented:
Hi, how would I go about saving each filename in this format (highlighted in blue):
![chrome_vBJ2LoC8Ux](https://github.com/user-attachments/assets/2fb670e8-779a-4172-830e-87be8e0fcfe2)

instead of this current default format?
![chrome_w0icE2BLT6](https://github.com/user-attachments/assets/94e36eec-de24-4928-abaf-b9fdd110ee84)

zWolfrost commented:
> Hi, how would I go about saving each filename in this format (highlighted in blue): ![chrome_vBJ2LoC8Ux](https://private-user-images.githubusercontent.com/177907820/356466546-2fb670e8-779a-4172-830e-87be8e0fcfe2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxODA0MDUsIm5iZiI6MTcyMzE4MDEwNSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NDY2NTQ2LTJmYjY3MGU4LTc3OWEtNDE3Mi04MzBlLTg3YmU4ZTBmY2ZlMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQwNTA4MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzJkMzY3OTg3MDBhNDMzMjgyYTExMjg2YzNhOWM5MWZkNjk5NmFhNDJmZGU0ZDNkZTMyYTZkNWMzNmVkNjU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-YyBg8BvvMrw3sac4InKPUshMYmms6qu4DAiITul0Tg)
> 
> instead of this current default format? ![chrome_w0icE2BLT6](https://private-user-images.githubusercontent.com/177907820/356466573-94e36eec-de24-4928-abaf-b9fdd110ee84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxODA0MDUsIm5iZiI6MTcyMzE4MDEwNSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NDY2NTczLTk0ZTM2ZWVjLWRlMjQtNDkyOC1hYmFmLWI5ZmRkMTEwZWU4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQwNTA4MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZjVmNWI0YzNjNDIwZGJmODhmYTEwMzg1NThmNzE5ODBmY2ZiNGQzZjhjZGU1YmU3MjU4ZTRmOGFiM2JmMWQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Hfuy1aEwKhU7RDhXQqN5ijlVd-Q29faJQAz5zhZMldA)

Hi, just pass the argument `-f "{filename}"` to the gallery-dl command or, alternatively, save it in the config (the same way you would do it for any other importer). I figured using the post ID to name them by default would be better, as you can trace the image back to its Facebook URL, but opinions are welcome.

lobt4 commented:
> > Hi, how would I go about saving each filename in this format (highlighted in blue): ![chrome_vBJ2LoC8Ux](https://private-user-images.githubusercontent.com/177907820/356466546-2fb670e8-779a-4172-830e-87be8e0fcfe2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxODA0MDUsIm5iZiI6MTcyMzE4MDEwNSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NDY2NTQ2LTJmYjY3MGU4LTc3OWEtNDE3Mi04MzBlLTg3YmU4ZTBmY2ZlMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQwNTA4MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzJkMzY3OTg3MDBhNDMzMjgyYTExMjg2YzNhOWM5MWZkNjk5NmFhNDJmZGU0ZDNkZTMyYTZkNWMzNmVkNjU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.-YyBg8BvvMrw3sac4InKPUshMYmms6qu4DAiITul0Tg)
> > instead of this current default format? ![chrome_w0icE2BLT6](https://private-user-images.githubusercontent.com/177907820/356466573-94e36eec-de24-4928-abaf-b9fdd110ee84.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxODA0MDUsIm5iZiI6MTcyMzE4MDEwNSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NDY2NTczLTk0ZTM2ZWVjLWRlMjQtNDkyOC1hYmFmLWI5ZmRkMTEwZWU4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQwNTA4MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZjVmNWI0YzNjNDIwZGJmODhmYTEwMzg1NThmNzE5ODBmY2ZiNGQzZjhjZGU1YmU3MjU4ZTRmOGFiM2JmMWQxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Hfuy1aEwKhU7RDhXQqN5ijlVd-Q29faJQAz5zhZMldA)
> 
> Hi, just pass the argument `-f "{filename}"` to the gallery-dl command or, alternatively, save it in the config (the same way you would do it for any other importer). I figured using the post ID to name them by default would be better, as you can trace the image back to its Facebook URL, but opinions are welcome.

Ah, gotcha. Straightforward for just one photo, but I'm looking to do this for an entire album of photos (for my own picky reasons). 

I'm thinking this will involve getting the "Open image in new tab" link, then grabbing the extended filename format from the link for each photo. Any idea how to go about doing this?

zWolfrost commented:
> Ah, gotcha. Straightforward for just one photo, but I'm looking to do this for an entire album of photos (for my own picky reasons).
> 
> I'm thinking this will involve getting the "Open image in new tab" link, then grabbing the extended filename format from the link for each photo. Any idea how to go about doing this?

I don't think I understand what you mean. The extractor is capable to extract the entire album of photos by itself, just copy the album URL (it should have this format `https://www.facebook.com/media/set/?set=a.10152716010956729&type=3`) and pass it to the command line, along with the parameter i suggested before. Remember to put the URL in double quotes (").

lobt4 commented:
> > Ah, gotcha. Straightforward for just one photo, but I'm looking to do this for an entire album of photos (for my own picky reasons).
> > I'm thinking this will involve getting the "Open image in new tab" link, then grabbing the extended filename format from the link for each photo. Any idea how to go about doing this?
> 
> I don't think I understand what you mean. The extractor is capable of extracting entire albums of photos by itself, just copy the album URL (it should have this format `https://www.facebook.com/media/set/?set=a.10152716010956729&type=3`) and pass it to the command line, along with the parameter i suggested before. Remember to put the URL in double quotes (").

My bad, I wasn't being clear. I'm looking to download from a `https://www.facebook.com/xxxxxxxxxx/photos` page (since not all of the photos are in albums) and instead of every photo being saved with this filename format:
![356466573-94e36eec-de24-4928-abaf-b9fdd110ee84](https://github.com/user-attachments/assets/cbe92fd8-c626-4899-9375-9a7bd74aa142)

I want the extractor to automatically save each photo with this filename format:
![356466546-2fb670e8-779a-4172-830e-87be8e0fcfe2](https://github.com/user-attachments/assets/406920ff-9ddf-4dc9-b77c-31f9e1788a3f)

zWolfrost commented:
> My bad, I wasn't being clear. I'm looking to download from a `https://www.facebook.com/xxxxxxxxxx/photos` page (since not all of the photos are in albums) and instead of every photo being saved with this filename format: ![356466573-94e36eec-de24-4928-abaf-b9fdd110ee84](https://private-user-images.githubusercontent.com/177907820/356682408-cbe92fd8-c626-4899-9375-9a7bd74aa142.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMjE3MzEsIm5iZiI6MTcyMzIyMTQzMSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NjgyNDA4LWNiZTkyZmQ4LWM2MjYtNDg5OS05Mzc1LTlhN2JkNzRhYTE0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQxNjM3MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNjYxYWE0ZDExNWY2MmI0NzAwM2YxZjRhMWNiNTVmYTRjNTBlNjI5ZWNhNDVkNjlhNzFkMDYwOGExZDlkYzNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9tmks7wBZxRAPz9liuHiBThW8JQRkklQMMNRfZo56zg)
> 
> I want the extractor to automatically save each photo with this filename format: ![356466546-2fb670e8-779a-4172-830e-87be8e0fcfe2](https://private-user-images.githubusercontent.com/177907820/356682589-406920ff-9ddf-4dc9-b77c-31f9e1788a3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMjE3MzEsIm5iZiI6MTcyMzIyMTQzMSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NjgyNTg5LTQwNjkyMGZmLTlkZGYtNGRjOS1iNzdjLTMxZjllMTc4OGEzZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQxNjM3MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMmVlOWIzYjliNDlhODRiYWI1MzI4NWU4MWE0NTZlZDU3ZDUzNjQ3OTFiMzdiZmYwMTc2Y2FmMjc4MjA4MjJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CMyIW2MZCrsM4wIi8cC6Z_YUwpC0cFKqeQR0LLfqlx4)

The extractor also supports downloading all photos from an account.
Just pass the account URL (e.g. `https://www.facebook.com/facebook/photos/`) and it will automatically extract all of them.

lobt4 commented:
> > My bad, I wasn't being clear. I'm looking to download from a `https://www.facebook.com/xxxxxxxxxx/photos` page (since not all of the photos are in albums) and instead of every photo being saved with this filename format: ![356466573-94e36eec-de24-4928-abaf-b9fdd110ee84](https://private-user-images.githubusercontent.com/177907820/356682408-cbe92fd8-c626-4899-9375-9a7bd74aa142.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMjE3MzEsIm5iZiI6MTcyMzIyMTQzMSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NjgyNDA4LWNiZTkyZmQ4LWM2MjYtNDg5OS05Mzc1LTlhN2JkNzRhYTE0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQxNjM3MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNjYxYWE0ZDExNWY2MmI0NzAwM2YxZjRhMWNiNTVmYTRjNTBlNjI5ZWNhNDVkNjlhNzFkMDYwOGExZDlkYzNkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9tmks7wBZxRAPz9liuHiBThW8JQRkklQMMNRfZo56zg)
> > I want the extractor to automatically save each photo with this filename format: ![356466546-2fb670e8-779a-4172-830e-87be8e0fcfe2](https://private-user-images.githubusercontent.com/177907820/356682589-406920ff-9ddf-4dc9-b77c-31f9e1788a3f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyMjE3MzEsIm5iZiI6MTcyMzIyMTQzMSwicGF0aCI6Ii8xNzc5MDc4MjAvMzU2NjgyNTg5LTQwNjkyMGZmLTlkZGYtNGRjOS1iNzdjLTMxZjllMTc4OGEzZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwOVQxNjM3MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMmVlOWIzYjliNDlhODRiYWI1MzI4NWU4MWE0NTZlZDU3ZDUzNjQ3OTFiMzdiZmYwMTc2Y2FmMjc4MjA4MjJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CMyIW2MZCrsM4wIi8cC6Z_YUwpC0cFKqeQR0LLfqlx4)
> 
> The extractor also supports downloading all photos from an account. Just pass the account URL (e.g. `https://www.facebook.com/facebook/photos/`) and it will automatically extract all of them.

I'm using the command `__main__.py https://www.facebook.com/xxxxxxxxxx/photos` and this is the error message I'm getting (can visit photo page on a browser)
![asd](https://github.com/user-attachments/assets/c87110f9-d04e-4c9b-bf54-afb1cd21202a)

lobt4 commented:
> Thanks for the report, that's definitely not supposed to happen. Is the profile you are trying to extract private? What is your OS language? (That has caused me problems in the past and if it's not English then i will start troubleshooting from there). Thanks for your patience.

No worries, I appreciate your help.
FB profile is public, OS language is English (US).

zWolfrost commented:
OK, are you using a VPN or anything that could change your country of residence? Some people country block their profile. Also, please consider [emailing](mailto:zwolfrost@gmail.com) me the profile you are trying to extract if you don't feel comfortable writing it here. I won't judge or anything, I've seen a lot of stuff. It would speed up the process a lot and i would really appreciate it. I really have no idea where to start right now...

lobt4 commented:
> OK, are you using a VPN or anything that could change your country of residence? Some people country block their profile. Also, is this only happening with this one profile? If yes, please consider [emailing](mailto:zwolfrost@gmail.com) me the profile you are trying to extract if you don't feel comfortable writing it here. I won't judge or anything, I've seen a lot of stuff. It would speed up the process a lot and i would really appreciate it. I really have no idea where to start right now...

Nope, not using a VPN and it seems like this is happening for every profile's "photos" page, but not for the photo albums that they've made.

I have a workaround where I use an extension to grab all the photo links on a "photos" page which I pass to gallery-dl in a text file to download the entire page. Though this circles back to my first question of trying to figure out how to automatically save all downloaded photos with the image ID (e.g. 344582_258993345612475_8286065_n.jpg) as its filename instead of the current default post ID.

zWolfrost commented:
> Nope, not using a VPN and it seems like this is happening for every profile's "photos" page, but not for the photo albums that they've made.
> 
> I have a workaround where I use an extension to grab all the photo links on a "photos" page which I pass to gallery-dl in a text file to download the entire page. Though this circles back to my first question of trying to figure out how to automatically save all downloaded photos with the image ID (e.g. 344582_258993345612475_8286065_n.jpg) as its filename instead of the current default post ID.

I assume that you mean "saving photos by their filename" without passing the argument that I've mentioned before every time (`-f "{filename}`)
In that case, you can configure gallery-dl to do exactly that every time by adding a `config.json` file in [those locations](https://github.com/mikf/gallery-dl?tab=readme-ov-file#locations) with something like this in it:
```
{
	"extractor": {
		"facebook": {
			"filename": "{filename}"
		}
	}
}
```
I will still definitely try to find the cause of the main issue.

lobt4 commented:
> > Nope, not using a VPN and it seems like this is happening for every profile's "photos" page, but not for the photo albums that they've made.
> > I have a workaround where I use an extension to grab all the photo links on a "photos" page which I pass to gallery-dl in a text file to download the entire page. Though this circles back to my first question of trying to figure out how to automatically save all downloaded photos with the image ID (e.g. 344582_258993345612475_8286065_n.jpg) as its filename instead of the current default post ID.
> 
> I assume that you mean "saving photos by their filename" without passing the argument that I've mentioned before every time (`-f "{filename}`) In that case, you can configure gallery-dl to do exactly that every time by adding a `config.json` file in [those locations](https://github.com/mikf/gallery-dl?tab=readme-ov-file#locations) with something like this in it:
> 
> ```
> {
> 	"extractor": {
> 		"facebook": {
> 			"filename": "{filename}"
> 		}
> 	}
> }
> ```
> 
> I will still definitely try to find the cause of the main issue.

Ah, well, that's embarrassing. I kept thinking I had to manually replace `{filename}` argument with the image filename inside the curly brackets instead of just using the argument as is. Not very smart of me.
The argument works just like you said and I'm able to automatically save photos with my preferred filenames.
Apologies for the confusion and thanks for your patience.

zWolfrost commented:
> Ah, well, that's embarrassing. I kept thinking I had to manually replace `{filename}` argument with the image filename inside the curly brackets instead of just using the argument as is. Not very smart of me. The argument works just like you said and I'm able to automatically save photos with my preferred filenames. Apologies for the confusion and thanks for your patience.

Don't worry, happens to the best of us. By the way, I've just fixed a thing that could have been the cause. Could you try again with the latest commit and let me know if it works instead?

lobt4 commented:
> Don't worry, happens to the best of us. By the way, I've just fixed a thing that could have been the cause. Could you try again with the latest commit and let me know if it works instead?

Errored out the same as last time, perhaps this is happening to just my machine for some reason:
![asd](https://github.com/user-attachments/assets/81b6b115-b895-4525-909f-8f3740c6a9c3)

AdBlocker69 commented:
Hi,
I just tried having yt-dlp handle the facebook video extraction and I'm getting this error:
![image](https://github.com/user-attachments/assets/6be279b5-7a14-4994-841d-f9b3c889ac02)
Might it be that it only searches for "youtube-dl" and therefore can't find it?

AdBlocker69 commented:
I don't have yt-dlp installed with Python, could that be the culprit? (I get `No module named yt_dlp`...)

zWolfrost commented:
Yeah, it most likely is. You should probably install yt-dlp with pip (follow a tutorial if you don't know how to, there's a lot of them). As far as i know there is no way to make gallery-dl use the yt-dlp executable instead of the python module.

AdBlocker69 commented:
Thanks, this way it works.
By default yt-dlp pulls the `best video` + `best audio` and merges them using FFmpeg.
In this case it would be "452530707751531v" + "1054256548546316a".
![image](https://github.com/user-attachments/assets/7668479a-70eb-4414-b4a3-8769a0531e3c)
How'd I go about telling yt-dlp to, for example, pull "hd" instead?
Because the following doesn't do the trick (normal way of selecting the yt-dlp module to use in gallery-dl):
![image](https://github.com/user-attachments/assets/87b8e04b-2387-48d1-a73e-02a2943004a9)

zWolfrost commented:
The reason why it's not working is because you should put that argument in the "downloader" options, not the "extractor" ones. Took me a bit 😅
```
{
	"extractor": {
		"#": "not here!"
	},

	"downloader": {
		"ytdl": {
			"format": "sd"
		}
	}
}
```

AdBlocker69 commented:
LOL, okay thanks, that works better now 😂...
Logically thinking it makes sense to be in "downloader" but I simply didn't think of that even being an option as I saw the same argument existing in the "extractor" options.
What is the one in there for then?

fireattack commented:
Sorry if this has been mentioned; but this seems to only be able to process the user, than the single post (`https://www.facebook.com/{username}/posts/{posthash}`

zWolfrost commented:
@fireattack I just fixed it. Let me know if now it works for you.

fireattack commented:
I now get error:

```
>python -m gallery_dl "https://www.facebook.com/joho.press.jp/posts/pfbid02mfFRpVkErLQxQ8cpD2f1hwXEVsFzK8kfNBKdK2Jndnx6AkmMQZuXhovwDgwvoDNil" 
[facebook][error] HttpError: '404 Not Found' for 'https://www.facebook.com/media/set/?set='
```

zWolfrost commented:
@fireattack The problem is that the extractor can't really get all the images in the posts. You would have a better chance copying the set id of one of those images, which is in their urls (in your case it's `pcb.1160563418981189`) and giving the set page to gallery-dl (`https://www.facebook.com/media/set/?set=pcb.1160563418981189`). Right now I'll see if I can do something about it (as well as fixing the error).

zWolfrost commented:
Right now the extractor should get the first set id it can find in the post and try to extract that set. Sometimes the set contains way more images than the post, but you could still quit it when it's done I guess. I could technically make it quit by itself when it has extracted the same number of images in the posts (36+4) but I'm not sure if I want to.

mikf commented:
I'd like to finally review and merge this PR to include it in the v1.28.0 release. Sorry for taking forever to get to this.

One of the general criticisms I have is your frequent use of `.*` and `.*?` in URL patterns. They should all be replaced with something more restrictive, at least `[^/?#]+` to not go over "URL boundaries". Are the URLs in `test/result/facebook.py` all that should be matched? If not, could you post a complete list of possible URLs and/or add them as only-matching entries to the results list?

mikf commented on docs/configuration.rst:
```diff
@@ -2376,6 +2376,30 @@ Description
+Description
+    Control video download behavior.
+
+    * ``true``: Extract and download video & audio separately.
+    * ``"ytdl"``: Let |ytdl| handle video extraction and download, and merge video & audio streams.
```
There should be an entry for `false` as well.

zWolfrost replied:
There isn't really a difference between false and true, should i just make the option a blank string when the video is handled by the extractor?

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+from .common import Extractor, Message
+from .. import text, exception
+
+BASE_PATTERN = r"(?:https?://)?.*?facebook\.com"
```
```suggestion
BASE_PATTERN = r"(?:https?://)?(?:\w+\.)?facebook\.com"
```
or something that matches whatever possible subdomains there are.

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+    set_url_fmt = root + "/media/set/?set={set_id}"
+    photo_url_fmt = root + "/photo/?fbid={photo_id}&set={set_id}"
+
+    def __init__(self, match):
+        Extractor.__init__(self, match)
+        self.match = match
+
```
```suggestion
```

`self.match` is already defined in the base `Extractor` class:
https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L53

----------------

All instances of `self.match.group(N)` in your could should generally be replaced with `self.groups[N-1]`
https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L54

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+        self.match = match
+
+    def _init(self):
+        self.session.headers["Accept"] = "text/html"
+        self.session.headers["Sec-Fetch-Mode"] = "navigate"
```
Might as well use a more complete set of headers from [here](https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L934-L968).

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+        self.session.headers["Sec-Fetch-Mode"] = "navigate"
+
+        self.fallback_retries = self.config("fallback-retries", 2)
+        self.sleep_429 = self.config("sleep-429", 5)
```
A `sleep_429` equivalent is already defined in the base Extractor class
https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L47
https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L333-L335

You can access it like [here](https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/extractor/common.py#L233-L237).

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+            txt.encode("utf-8").decode("unicode_escape")
+        ).replace("\\/", "/")
+
+    @staticmethod
+    def item_filename_handle(item):
+        item["filename"] = item.get("filename") or ""
+        if "." in item["filename"]:
+            item["name"], item["extension"] = item["filename"].rsplit(".", 1)
+        else:
+            item["name"] = item["filename"]
+            item["extension"] = ""
```
Filenames are usually handled by [`text.nameext_from_url()`](https://github.com/mikf/gallery-dl/blob/d1ad97ae0c1417bbcabb84fba13ecd514ba11223/gallery_dl/text.py#L88)

Also, I prefer using `str.partiton()` over `str.split()` when splitting in two, but this is not that important.

zWolfrost replied:
Should I just ditch the "name" keyword then?

mikf replied:
Yes. A file called `name.ext` should have `name` as `filename` and `ext` as `extension`. At least that's the usual convention.

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+                LEFT_OFF_TXT
+            )
+
+        if '{"__dr":"CometErrorRoot.react"}' in res.text:
```
```suggestion
        if b'{"__dr":"CometErrorRoot.react"}' in res.content:
```

Avoid accessing ´res.text` more than once. It's a quite costly operation.

zWolfrost commented:
The URLs in the test script should be the complete list of all of them. The only exception is the `facebook.FacebookProfileExtractor`, which is kind of a "fallback" for the others, and would work for all instances of a facebook link that starts with "`https://www.facebook.com/[username]/...`" (and also "profile.php\?id=" links which I added now in the tests). I made it like this because there are many account URLs that one would expect to work and extract the whole account when passed to gallery-dl. Of course i can change this if you want to, along with the other things.
Now that we're at it, do you happen to know why the archive ids doesn't seem to have the "extension" part of the format in the name? (thus failing one of the tests?)

mikf commented on gallery_dl/extractor/facebook.py:
```diff
@@ -0,0 +1,454 @@
+            "\n" + url + "&setextract"
+        )
+
+        res = self.request(url, *args, **kwargs)
```
```suggestion
        res = self.request(url, **kwargs)
```

`Extractor.request()` has only `url` as positional argument.

zWolfrost commented:
I have written "most of the requested changes" in the commit message because I can't really replace the ".*" and ".*?" URL patterns with "[^/?#]+" most of the times because they are often intended to skip over URL paths. I have changed the other stuff.

zWolfrost commented:
I forgot to tell you about the "&setextract" function, where if a photo URL ends with "&setextract" the extractor will instead go through the whole set starting from that photo. This is to avoid having to start from the beginning after being temporarily banned. I have added the remaining tests and adjusted the concerned patterns.

MikeRich88 commented:
I'd like to test this, but the nightly builds are no good for Intel macs :(

Is there some reason it's not being built as universal? You can build one arm64 and one x86_64 and then `lipo` them together.

>QuinnHex:Desktop mike$ chmod +x /Users/mike/Downloads/gallery-dl_macos 
>QuinnHex:Desktop mike$ /Users/mike/Downloads/gallery-dl_macos 
>-bash: /Users/mike/Downloads/gallery-dl_macos: Bad CPU type in executable

MikeRich88 commented:
Was able to download a friend's photos successfully, although they only have 6 photos ;)

zWolfrost commented:
@MikeRich88 I'm glad to hear that it works for you. To answer to your question, the extractor has to make a request to a webpage before every image (+one at the beginning) that are **really** expensive, to get the necessary data. Just one request is ~1.5MB so yeah, unless Facebook optimizes his pages the delay will always be there.
